### PR TITLE
feat: item variant sync

### PR DESCRIPTION
**Pull request by rtdany10**
https://user-images.githubusercontent.com/52111700/196864889-f6e568ba-0aea-4839-9909-8e94483010f9.mp4

### CODE REVIEW

**Comment by rtdany10** [comment]
ToDo:

- [ ] Disallow adding new attribute to item template after variants are created.
- [ ] Do not sync items with more than 3 attributes.

**Comment by rtdany10** [comment]
@ankush :)

**Comment by rtdany10** [comment]
@ankush :grimacing:

**Review by ankush — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
Inventory sync for created variant doesn't seem to work well.

I imported item with 4 color variants and added stock for black and green
image


Shopify shows stock for red


image

**Comment by ankush** [comment]
4 attribute items are silently uploading partial details and failing! This is not good. 

We should just warn that it won't be synced and don't sync it.

ERPNext item:

image



Shopify item

image

File: ecommerce_integrations/shopify/product.py [hunk 1]
```diff
@@ -342,51 +342,125 @@ def upload_erpnext_item(doc, method=None):
...
-			ecom_item.insert()
+			ecom_items = list(set([item, template_item]))
+			for d in ecom_items:
+				ecom_item = frappe.get_doc(
+					{
+						"doctype": "Ecommerce Item",
+						"erpnext_item_code": d.name,
+						"integration": MODULE_NAME,
+						"integration_item_code": str(product.id),
+						"variant_id": "" if d.has_variants else str(product.variants[0].id),
```
Comment (ankush) [inline ecommerce_integrations/shopify/product.py:409]: They ALL have same variant ID that's why inventory sync behaviour is buggy

image



At least test once!

**Review by ankush — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
TODO:
- [ ] Fix setting for making products active by default
- [ ] Fix variant id mapping and perhaps add a test for this logic, the other way around has tests https://github.com/frappe/ecommerce_integrations/blob/develop/ecommerce_integrations/shopify/tests/test_product.py#L28-L65
- [ ] test manually again

File: ecommerce_integrations/shopify/product.py (hunk 1 above)
Comment (rtdany10) [reply ecommerce_integrations/shopify/product.py:409]: Shopify seem to not have a fixed order for the variants. After each `product.save()`, the order of the list seems to mutate.
I had limited number of attribute values and unfortunately, new variants came as the last ones when I was writing the logic and I decided to go with `-1` index.

Now that I check with more data, it breaks. I have written loop to check for matching conditions and then set the variant ID.

File: ecommerce_integrations/shopify/product.py [hunk 2]
```diff
@@ -342,51 +342,125 @@ def upload_erpnext_item(doc, method=None):
 	if frappe.flags.in_import:
 		return
 
-	if doc.has_variants or doc.variant_of:
-		msgprint(_("Item with variants or template items can not be uploaded to Shopify."))
+	if item.has_variants:
+		# msgprint(_("Template items can not be uploaded to Shopify."))
```
[the remark this answers is not in the document]
Comment (rtdany10) [reply ecommerce_integrations/shopify/product.py]: Have replaced it with a message.

**Comment by rtdany10** [comment]
> 4 attribute items are silently uploading partial details and failing! This is not good.
> 
> We should just warn that it won't be synced and don't sync it.

> dont upload for >3 attributes

The reason I allowed it was because Shopify allows 3 attributes _by default_, but can be increased using third party apps. Ref: https://help.shopify.com/en/manual/products/variants/add-variants#considerations-for-adding-variants

Anyway, I have limited it to 3 for now.

File: ecommerce_integrations/shopify/doctype/shopify_setting/shopify_setting.json [hunk 3]
```diff
@@ -327,12 +330,29 @@
...
+  {
+   "default": "0",
+   "description": "Caution: Only 3 attributes will be accepted by Shopify",
+   "fieldname": "upload_variants_as_items",
+   "fieldtype": "Check",
+   "label": "Upload ERPNext Variants as Shopify Items"
+  },
+  {
+   "default": "1",
+   "fieldname": "sync_new_item_as_draft",
```
[the remark this answers is not in the document]
Comment (rtdany10) [reply ecommerce_integrations/shopify/doctype/shopify_setting/shopify_setting.json]: Done

**Comment by rtdany10** [comment]
Stock issue has been resolved after correctly setting the variant ID. Have tested it locally and didn't find any issue.
Unsure of how to write a test tho.

@ankush

**Comment by ankush** [comment]
@rtdany10 we can write test that at-least verifies mapping of attributes is done correctly? 

Split the function from that hook... will need lil bit of refactoring the code to make it easily testable.

**Comment by rtdany10** [comment]
@ankush Have added test case to check mapping

**Comment by rtdany10** [comment]
@ankush

**Pull request merged**